src/process/base/ZLibBasedFASTQProcessorBaseClass.cpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2019 DKFZ - ODCF
+ *
+ * Distributed under the MIT License (license terms are at https://github.com/dkfz-odcf/FastqIndEx/blob/master/LICENSE.txt).
+ */
+
+#include "process/io/PathSource.h"
+#include "process/extract/Extractor.h"
+#include "ZLibBasedFASTQProcessorBaseClass.h"
+#include <cstdio>
+#include <experimental/filesystem>
+#include <iostream>
+#include <unistd.h>
+#include <zlib.h>
+
+ZLibBasedFASTQProcessorBaseClass::ZLibBasedFASTQProcessorBaseClass(
+        shared_ptr<Source> fastq,
+        shared_ptr<Source> index,
+        const bool enableDebugging) :
+        enableDebugging(enableDebugging) {
+
+    // clang-tidy will complain, that zStream is not initialized by this constructor. This is done in another step.
+    fastqFile = move(fastq);
+    inputIndexFile = move(index);
+}
+
+ZLibBasedFASTQProcessorBaseClass::ZLibBasedFASTQProcessorBaseClass(
+        shared_ptr<Source> fastq,
+        shared_ptr<Sink> index,
+        const bool enableDebugging) :
+        enableDebugging(enableDebugging) {
+
+    // clang-tidy will complain, that zStream is not initialized by this constructor. This is done in another step.
+    fastqFile = move(fastq);
+    outputIndexFile = move(index);
+}
+
+bool ZLibBasedFASTQProcessorBaseClass::initializeZStream(int mode) {
+    zStream.zalloc = nullptr;
+    zStream.zfree = nullptr;
+    zStream.opaque = nullptr;
+    zStream.next_in = nullptr;
+    zStream.avail_in = 0;
+    zStream.avail_out = 0;
+
+    zlibResult = inflateInit2(&zStream, mode);
+    if (zlibResult != Z_OK) {
+        addErrorMessage("The zlib stream could not be initialized.");
+        return false;
+    }
+
+    return true;
+}
+
+bool ZLibBasedFASTQProcessorBaseClass::readCompressedDataFromSource() {
+    /* get some compressed data from input file */
+    int result = this->fastqFile->read(input, CHUNK_SIZE);
+
+    if (result == -1) {
+        this->addErrorMessage("There was an error during fread.");
```

Comment:
Is there actually a way to say what kind of error? E.g. pipeerror? IO-error? If so, this information should be added to the error message, because that enormeously helps error diagnosis when using the tool.